### ISSUE TITLE
Possible spam from the password reset functionality

### ISSUE THREAD

**Issue by intel-xeon** [keycloak/keycloak#24914]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

authentication

### Describe the bug

This bug impacts the password reset functionality and involves misuse of the password to create endless password reset requests in order to clog up a potential victim's inbox.

### Version

Confirmed on 22.0.5

### Expected behavior

The link used to abuse the password reset functionality must be revoked once it has been used and cannot be used for another reset. It is also advisable to set a minimum time (e.g. 5 minutes) within which it will not be possible to carry out a new password reset for a user

### Actual behavior

The link used to abuse the password reset functionality is NOT revoked once used and can be used to send multiple reset requests. 

### How to Reproduce?

Here are the details of the discovery:

First, go to the form that handles the password reset.

Once on the form, simply take the URL that appears above the address bar, enter a valid reset e-mail and click "submit" to start the password reset process. 
 
Once "submit" has been clicked, we can take the previously taken URL and call it up several times without any control in order to send multiple password reset messages to clog up the previously chosen e-mail address.
 

### Anything else?

CWE: https://cwe.mitre.org/data/definitions/640.html
CAPEC: https://capec.mitre.org/data/definitions/212.html

**Comment by mposolda** [comment]
I am not sure what to do with this one. I can understand why someone can consider this the issue...

On the other hand, some opinion on this from @rmartinc :
You can always spam a user email if you know it. What's the difference between reusing the link or just start a new process in a new (incognito) window? I have done the same in a personal google account (but using back button in the browser) and I sent 6 emails to my recovery email.


I don't know if we can somehow improve and introduce some kind of rate limit? But might be non-trivial effort...

In theory, we can re-use brute-force. If sending email during "Forget password" flow is considered as "failed login attempt", then it ensures that there won't be easily possible to do something like DoS attack by flooding your mailbox with huge amount of emails.

As actual password-reset will mark brute-force as "success" and delete it. But not 100% sure if using Brute-Force is a workaround?

**Comment by mposolda** [comment]
Adding to Backlog with label "Help wanted". Keycloak team likely won't have time to look at this in the near future, but anyone is welcome to send PR with improving this. Possibly using brute-force as mentioned above might be an easiest way to go.

**Comment by lexcao** [comment]
Hi @mposolda 
I have an idea by adding the reCaptcha like registration form does to make the link be used only once.
WDYT?

**Comment by pratkpranav** [comment]
Any updates on this?

**Comment by ajmas** [comment]
We recently ran into this issue because a user what spamming the reset button. The user's host was Yahoo/

The response we were seeing in the SendGrid logs:

```
Email was deferred due to the following reason(s): [IPs were throttled by recipient server]
```

**Comment by idNoRD** [comment]
As a temporary solution https://github.com/idNoRD/keycloak-spi-limit-resend-email
